components/places/src/storage/history.rs

```diff
@@ -1019,8 +1019,20 @@ pub mod history_sync {
     }
 
     pub fn apply_synced_deletion(db: &PlacesDb, guid: &SyncGuid) -> Result<()> {
+        // First we delete any visits for the page
+        // because it's possible the moz_places foreign_count is not 0
+        // and thus the moz_places entry won't be deleted.
         db.execute_cached(
-            "DELETE FROM moz_places WHERE guid = :guid",
+            "DELETE FROM moz_historyvisits
+              WHERE place_id IN (
+                  SELECT id
+                  FROM moz_places
+                  WHERE guid = :guid
```

Comment:
I thought of refactoring the guid scan here so we do it only once.. but we have unique indexes on the guid so that wouldn't have made too much of a difference in practice (but I could be corrected!)

Reply:
I think SQLite's optimizer will do the right thing here! `(SELECT id FROM moz_places WHERE guid = :guid)` isn't a correlated subquery, so it should be run just once to build the `IN` list—which, thanks to the unique constraint on `moz_places.guid`, will contain just one entry at most—and then use the `placedateindex` on `moz_historyvisits` to search and delete all matching rows.